### PR TITLE
Loki: Add support for ad-hoc filtering in dashboard

### LINKED ISSUE #35108: Loki: Add Ad-hoc filtering in dashboard
**What would you like to be added**:

Support ad hoc filtering in dashboard such as https://play.grafana.org/d/000000014/elasticsearch-metrics?orgId=1. 
image

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:
This PR adds ad-hoc filtering support for Loki - such as Support ad hoc filtering in dashboard such as https://play.grafana.org/d/000000014/elasticsearch-metrics?orgId=1.
**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/35108

**Special notes for your reviewer**:
1. Have active loki data source
2. Go to dashboards
3. Go to variables -> Create variable
4. Choose ad hoc filtering + choose loki as a data source
5. Create panel with loki query
6. Choose new label + value from dropdowns and see if it was correctly added to the query

### CODE REVIEW

File: docs/sources/datasources/loki.md [hunk 1]
```diff
@@ -154,6 +154,9 @@ provides the following functions you can use in the `Query` input field.
 | `label_values(label)`                      | Returns a list of label values for the `label`.                                      |
 | `label_values(log stream selector, label)` | Returns a list of label values for the `label` in the specified `log stream selector`.|
 
+### Ad hoc filters variable
+Loki supports the special ad hoc filters variable type. It allows you to specify any number of label/value filters on the fly. These filters are automatically applied to all your Loki queries.
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/loki.md:158]: is this specific to Loki only, or other data sources can use this variable? 

As a best practice, better to not use "special" in product docs.

**Review by achatterjee-grafana — CHANGES_REQUESTED** [review]

File: docs/sources/datasources/loki.md (hunk 1 above)
Comment (ivanahuckova) [reply docs/sources/datasources/loki.md:158]: `Ad hoc filter variables only work with Prometheus, Loki, InfluxDB and Elasticsearch data sources.` I used the same copy as it is in Prometheus https://grafana.com/docs/grafana/latest/datasources/prometheus/#ad-hoc-filters-variable. We can remove `special` from both places.

File: docs/sources/variables/variable-types/add-ad-hoc-filters.md [hunk 2]
```diff
@@ -8,7 +8,7 @@ weight = 700
 
 _Ad hoc filters_ allow you to add key/value filters that are automatically added to all metric queries that use the specified data source. Unlike other variables, you do not use ad hoc filters in queries. Instead, you use ad hoc filters to write filters for existing queries.
 
-> **Note:** Ad hoc filter variables only work with  InfluxDB, Prometheus, and Elasticsearch data sources.
+> **Note:** Ad hoc filter variables only work with Prometheus, Loki, InfluxDB and Elasticsearch data sources.
```
Comment (achatterjee-grafana) [inline docs/sources/variables/variable-types/add-ad-hoc-filters.md:11]:
```suggestion
> **Note:** Ad hoc filter variables only work with Prometheus, Loki, InfluxDB, and Elasticsearch data sources.
```

**Review by achatterjee-grafana — COMMENTED** [review]
Great! Thank you. Minor copy edit for the note.

**Review by achatterjee-grafana — APPROVED** [review]

**Comment by achatterjee-grafana** [comment]
Need backport to 8.0.x doc set.

File: public/app/plugins/datasource/loki/datasource.ts [hunk 3]
```diff
@@ -415,6 +419,19 @@ export class LokiDatasource extends DataSourceApi<LokiQuery, LokiOptions> {
     return Math.ceil(date.valueOf() * 1e6);
   }
 
+  // By implementing getTagKeys and getTagValues we add ad-hoc filtters functionality
+  async getTagKeys() {
+    const params = this.getTimeRangeParams();
```
[the remark this answers is not in the document]
Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/datasource.ts]: Good catch! I have refactored this to 
```
 async getTagKeys() {
    return await this.labelNamesQuery();
  }

  async getTagValues(options: any = {}) {
    return await this.labelValuesQuery(options.key);
  }
```

**Review by zoltanbedi — APPROVED** [review]

**Comment by ivanahuckova** [comment]
>Need backport to 8.0.x doc set.

@achatterjee-grafana This will be part of 8.1, so it shouldn't need to be backported.

**Pull request merged**